src/qiboml/operations/expectation.py

```diff
@@ -0,0 +1,138 @@
+            values.
+        differentiation_rule (Optional[callable]): the chosen differentiation
+            rule. It can be selected among the methods implemented in
+            ``qiboml.differentiation``.
+    """
+
+    # read the frontend user choice
+    frontend = observable.backend
+
+    kwargs = dict(
+        observable=observable,
+        circuit=circuit,
+        initial_state=initial_state,
+        nshots=nshots,
+        backend=backend,
+        differentiation_rule=differentiation_rule,
+    )
+    if isinstance(frontend, TensorflowBackend):
+        return _with_tf(**kwargs)
+
+    raise_error(
+        NotImplementedError,
+        "Only tensorflow automatic differentiation is supported at this moment.",
+    )
+
+
+def _exact(observable, circuit, initial_state, exec_backend):
+    """Helper function to compute exact expectation values."""
+    return observable.expectation(
+        exec_backend.execute_circuit(
+            circuit=circuit, initial_state=initial_state
+        ).state()
+    )
+
+
+def _with_shots(observable, circuit, initial_state, nshots, exec_backend):
+    """Helper function to compute expectation values from samples."""
+    return exec_backend.execute_circuit(
+        circuit=circuit, initial_state=initial_state, nshots=nshots
+    ).expectation_from_samples(observable)
+
+
+def _with_tf(
+    observable,
+    circuit,
+    initial_state,
+    nshots,
+    backend,
+    differentiation_rule,
+):
+    """
+    Compute expectation sample integrating the custom differentiation rule with
+    TensorFlow's automatic differentiation.
+    """
+    import tensorflow as tf  # pylint: disable=import-error
+
+    params = circuit.get_parameters()
+    nparams = len(params)
+
+    exec_backend = construct_backend(backend)
+
+    @tf.custom_gradient
+    def _expectation(params):
+        params = tf.Variable(params)
+
+        def grad(upstream):
+            gradients = []
+            for p in range(nparams):
+                gradients.append(
+                    upstream
+                    * differentiation_rule(
+                        circuit=circuit,
+                        hamiltonian=observable,
+                        parameter_index=p,
+                        initial_state=initial_state,
+                        nshots=nshots,
+                        backend=backend,
+                    )
+                )
+            return gradients
+
+        if nshots is None:
+            expval = _exact(observable, circuit, initial_state, exec_backend)
+        else:
+            expval = _with_shots(
+                observable, circuit, initial_state, nshots, exec_backend
+            )
+
+        return expval, grad
+
+    if differentiation_rule is not None:
+        return _expectation(params)
+
+    elif nshots is None:
+        return _exact(observable, circuit, initial_state, exec_backend)
+
+    else:
+        return _with_shots(observable, circuit, initial_state, nshots, exec_backend)
```

Comment:
Since they are  not specific to TensorFlow, you could just return the `_expectation` closure and move these into the general `expectation()`

Reply:
Done, thanks! :)